lib/descendDisplay.ml

```diff
@@ -1282,6 +1306,573 @@ let print_aboville conf base max_level p =
+
+(* hbar over xn - x1 columns. First and last are left/right and 50% *)
+let td_hbar x1 xn =
+  match xn - x1 with
+  | 0 -> [ (1, CenterA, TDnothing) ]
+  | 1 -> [ (1, CenterA, TDhr CenterA) ]
+  | _ ->
+      [ (1, LeftA, TDhr LeftA) ]
+      @ [ (xn - x1 - 1, CenterA, TDhr CenterA) ]
+      @ [ (1, RightA, TDhr RightA) ]
+
+(* regular cell, centered, with text as content (may contain |<br>) *)
+let td_cell cols align ip text flags =
+  [ (cols, align, TDitem (ip, Adef.safe text, flags)) ]
+
+(* tdal is   (int   *    list      ) list           *)
+(*           (lastx      list of td) list of rows   *)
+
+(* add elem to row ir. Update lastx *)
+let tdal_add tdal ir elem nx =
+  let tdal =
+    let rec loop tdal new_tdal ir elem nx =
+      match tdal with
+      | [] -> new_tdal
+      | (x, row) :: tdal ->
+          if ir = 0 then
+            let new_row = (nx, row @ elem) in
+            new_row :: loop tdal new_tdal (ir - 1) elem nx
+          else (x, row) :: loop tdal new_tdal (ir - 1) elem nx
+    in
+    loop tdal [] ir elem nx
+  in
+  tdal
+
+let get_bd_td_prop conf =
+  let bd = match Util.p_getint conf.env "bd" with Some x -> x | None -> 0 in
+  let td_prop =
+    match Util.p_getenv conf.env "td" with
+    | Some x -> " " ^ x
+    | _ -> (
+        match Util.p_getenv conf.env "color" with
+        | None | Some "" -> ""
+        | Some x -> " class=\"" ^ x ^ "\"")
+  in
+  (bd, td_prop)
+
+let reference conf base p s =
+  let cgl =
+    match Util.p_getenv conf.env "cgl" with Some "on" -> true | _ -> false
+  in
+  let iper = get_iper p in
+  if is_hidden p || cgl then s
+  else
+    Printf.sprintf
+      {|<a href="%s%s" id="i%s" class="normal_anchor" title="%s">%s</a>|}
+      (commd conf :> string)
+      (acces conf base p :> string)
+      (string_of_iper iper)
+      (Utf8.capitalize_fst (Util.transl conf "open individual page"))
+      s
+
+let get_text conf base p filler img cgl =
+  let bd, td_prop = get_bd_td_prop conf in
+  let auth = authorized_age conf base p in
+  let txt = person_title_text conf base p in
+  let txt =
+    if cgl then (txt :> string) else reference conf base p (txt :> string)
+  in
+  let txt =
+    if auth then txt ^ (DateDisplay.short_dates_text conf base p :> string)
+    else txt
+  in
+  let vbar_image =
+    if filler && img then
+      Printf.sprintf
+        "<div class=\"flex-grow-1\"><img src=\"%sm=IM&v=vbar.jpg\" \
+         class=\"img-fluid\"></div>\n"
+        (commd conf :> string)
+    else ""
+  in
+  let has_image = Image.get_portrait conf base p |> Option.is_some in
+  let txt =
+    txt
+    ^
+    if has_image && img then (DagDisplay.image_txt conf base p :> string)
+    else vbar_image
+  in
+  let txt =
+    if bd > 0 || td_prop <> "" then
+      Printf.sprintf
+        "<table style=\"border: %dpx solid\"><tr><td \
+         align=\"center\"%s>%s</td></tr></table>"
+        bd td_prop txt
+    else txt
+  in
+  txt
+
+let lastx tdal ir =
+  let x, _ = List.nth tdal ir in
```

Comment:
if we need to access the nth element of `tdal` a lot, maybe an array is a better data structure